datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -441,11 +442,11 @@ fn push_down_all_join(
 
     // Extract from OR clause, generate new predicates for both side of join if possible.
     // We only track the unpushable predicates above.
-    if left_preserved {
+    if on_left_preserved {
         left_push.extend(extract_or_clauses_for_join(&keep_predicates, left_schema));
         left_push.extend(extract_or_clauses_for_join(&join_conditions, left_schema));
     }
-    if right_preserved {
+    if on_right_preserved {
```

Comment:
We have discussed the conditions, filter predicates and on expression can be pushed down with @ozankabak .
According to us, correct handling seems to be 
```
    if left_preserved {
        left_push.extend(extract_or_clauses_for_join(&keep_predicates, left_schema));
    }
    if on_left_preserved {
        left_push.extend(extract_or_clauses_for_join(&join_conditions, left_schema));
    }
    if right_preserved {
        right_push.extend(extract_or_clauses_for_join(&keep_predicates, right_schema));
    }
    if on_right_preserved {
        right_push.extend(extract_or_clauses_for_join(&join_conditions, right_schema));
    }
```
where, for predicates and join conditions flag used is different. What do you think?

Reply:
I am not sure I understand what is different about your suggestion.  

My understanding of the code is that to breaks the predicates into:
* `keep_predicates` predicates that must be left as a filter (above the join)
* `join_conditions` predicates that must be left in the join (e.g. `ON` clause)
* `left_push` predicates that can be pushed down to the left input
* `right_push` predicates that can be pushed down to the right input

I think the code, as written, does push predicates from the ON clause down correctly when possible. 

So in other words, I am not sure what changes you are suggesting

Reply:
We could be missing something but it seems like `keep_predicates` and `join_conditions` should be using `preserved` vs. `on_preserved` (respectively) since the former applies as a post-join filter.

Reply:
For left join, `left_preserved=true` and `on_left_preserved=false`.
In current version of the code, nothing from `keep_predicates` and `join_conditions` can be inserted to the `left_push` since `on_left_preserved` flag is `false`. However,  if we use the flag `left_preserved` to decide whether we can insert  something from `keep_predicates`, into `left_push`. We could have inserted additional conditions to the `left_push`. In this sense, current code works correctly, however behaves suboptimal in some cases.

Reply:
Actually we wrongly mix join filter predicates and filter predicates into `join_conditions`. I separate them now so only `on_filter_join_conditions` will be considered with `on_preserved`.

Reply:
Thank you for the clarifications


> In current version of the code, nothing from keep_predicates and join_conditions can be inserted to the left_push since on_left_preserved flag is false

I agree, but I think this is required for correctness. At least for `join_conditions` it is important not to push them down to the preserved side 

For example, in this query  I am pretty sure it is not correct to push the predicate on `A.a` below the join (the intuition is that all rows from `A` have to have at least one row out of the join). I believe this is what this PR fixes. 

```SQL
SELECT .. FROM A LEFT JOIN B ON (A.a > 5)
```

I think *is* correct to push predicates on B below (as any non matching rows would have been filtered in the join anyways)

```SQL
SELECT .. FROM A LEFT JOIN B ON (B.b > 5)
```

> We could have inserted additional conditions to the left_push. In this sense, current code works correctly, however behaves suboptimal in some cases.

Can you help me write an example? I can add it to this PR and then we can handle improving the case in a future PR.

Reply:
I tried to add a test case with it.

Reply:
I think we found one example -- @mustafasrepo will post shortly